network/peer.go

```diff
@@ -50,7 +60,13 @@ type peer struct {
 
 	// ip may or may not be set when the peer is first started. is only modified
 	// on the connection's reader routine.
-	ip     utils.IPDesc
+	ip utils.IPDesc
+
+	// aliases is a list of IPs other than [ip] that we have connected to
+	// this peer at.
+	aliases []alias
+
+	// ipLock must be held when accessing either [ip] or [aliases].
 	ipLock sync.RWMutex
```

Comment:
There's a pre-existing comment on the network stateLock that says it should never be held while grabbing a peer lock. I think that's in reference to the other lock, not the ipLock, but could you add a comment to clarify?

Reply:
Strangely enough, that rule is already not followed in `disconnected` as `p.getIP` acquires the "peer lock" while we hold `stateLock`: https://github.com/ava-labs/avalanchego/blob/8be88a342fced5522cd503b72f49aae450eea863/network/network.go#L1130

Reply:
In fact, the PR that added this comment (#508) added the code that bypassed the rule expressed in that comment.

Reply:
https://github.com/ava-labs/avalanchego/pull/508/files#diff-96a7f5c352cd3d2d1ee3c7c55ed95e668c6edb28a4fa337e6c60270471bdab61R1082-R1085

Reply:
Yup, I believe the comment is in reference to the other peer lock, @StephenButtolph could you confirm?

Reply:
Golang will panic if two people update the values at the same time.
ipLock locks the getting or setting of the IP..

That's why the ipLock was created.

The PR in reference makes things parallel which means you have to careful when you update internals of the peer.